libshpool/src/config.rs

```diff
@@ -12,36 +12,120 @@
+pub struct Manager {
+    /// The config value.
+    config: Arc<RwLock<Config>>,
+    watcher: Option<Arc<notify::RecommendedWatcher>>,
+}
+
+impl Manager {
+    // Create a new config manager.
+    pub fn new(config_file: Option<&str>) -> anyhow::Result<Self> {
         let user_info = user::info()?;
-        let mut config_path = PathBuf::from(user_info.home_dir);
-        config_path.push(".config");
-        config_path.push("shpool");
-        config_path.push("config.toml");
-        if config_path.exists() {
-            let config_str = fs::read_to_string(config_path).context("reading config toml (2)")?;
-            config = toml::from_str(&config_str).context("parsing config file (2)")?;
+        let mut default_config_path = PathBuf::from(user_info.home_dir);
+
+        let mut config = Config::default();
+        let config_path = if let Some(config_path) = config_file {
+            info!("parsing explicitly passed in config ({})", config_path);
+            let config_str = fs::read_to_string(config_path).context("reading config toml (1)")?;
+            config = toml::from_str(&config_str).context("parsing config file (1)")?;
+
+            Some(String::from(config_path))
+        } else {
+            default_config_path.push(".config");
+            default_config_path.push("shpool");
+            default_config_path.push("config.toml");
+            if default_config_path.exists() {
+                let config_str =
+                    fs::read_to_string(&default_config_path).context("reading config toml (2)")?;
+                config = toml::from_str(&config_str).context("parsing config file (2)")?;
+
+                default_config_path.clone().to_str().map(String::from)
+            } else {
+                None
+            }
+        };
+
+        let mut manager = Manager { config: Arc::new(RwLock::new(config)), watcher: None };
+
+        if let Some(watch_path) = config_path {
+            let config_slot = Arc::clone(&manager.config);
+            let reload_path = watch_path.clone();
+            let mut watcher = notify::recommended_watcher(move |res| match res {
+                Ok(event) => {
+                    info!("config file modify event: {:?}", event);
+
+                    let config_str = match fs::read_to_string(&reload_path) {
+                        Ok(s) => s,
+                        Err(e) => {
+                            warn!("error reading config file: {:?}", e);
+                            return;
+                        }
+                    };
+
+                    let config = match toml::from_str(&config_str) {
+                        Ok(c) => c,
+                        Err(e) => {
+                            warn!("error parsing config file: {:?}", e);
+                            return;
+                        }
+                    };
+
+                    let mut manager_config = config_slot.write().unwrap();
```

Comment:
Yes, this can panic, but it will only panic on a poisoned mutex (well technically rwlock). In the case of a poisoned mutex, there really isn't anything reasonable you can do to recover and data might be corrupted. It is a pretty rare condition that never really happens in practice. The recommended practice is to  unwrap since unwinding the stack for the current thread is pretty much the only safe thing to do.

Reply:
sounds reasonable, thanks for explaining!